app/Models/User.php

```diff
@@ -170,10 +171,15 @@ public function devices()
         if ($this->hasGlobalRead()) {
             return Device::query();
         } else {
-            return $this->belongsToMany('App\Models\Device', 'devices_perms', 'user_id', 'device_id');
+            return Device::HasAccess($this);
```

Comment:
This needs to remain a query

Reply:
What do you mean? the existing code is fine.

Reply:
I think you are trying to put complexity at the wrong level.
If this isn't a query, then we can't do a lot of relationship based queries.
If you want a function that gets all devices a user has access to using the cache, create a new function.